nbt-utils/nbt_utils.cpp

```diff
@@ -270,6 +276,422 @@ wr_payload(LongArrayTag) {
+ 
+#define rd_snbt_suffix(tagCls, readType, suffixUpper, suffixLower) \
+  rd_snbt(tagCls) { \
+    std::string buf; \
+    char c; \
+    while (stream.get(c)) { \
+      if (c == suffixUpper || c == suffixLower) \
+        break; \
+      if (c >= '0' && c <= '9' || c == '.' || c == '-') \
+        buf.push_back(c); \
+      else \
+        return false; \
+    } \
+    std::stringstream bufStream(buf); \
+    readType value_; \
+    bufStream >> value_; \
+    if (bufStream.get() != EOF) \
+      return false; \
+    if (!stream) \
+      return false; \
+    value = (decltype(value)) value_; \
+    return true; \
+  }
+
+
+rd_snbt_suffix(ByteTag, int, 'B', 'b');
+wr_snbt(ByteTag) { stream << (int) value << 'b'; }  // C++ cannot distinguish byte from char
+
+rd_snbt_suffix(ShortTag, int16_t, 'S', 's');
+wr_snbt(ShortTag) { stream << value << 's'; }
+
+rd_snbt(IntTag) { stream >> value; return !!stream; }  // JavaScript moment
+wr_snbt(IntTag) { stream << value; }
+
+rd_snbt_suffix(LongTag, int64_t, 'L', 'l');
+wr_snbt(LongTag) { stream << value << 'l'; }
+
+rd_snbt_suffix(FloatTag, float, 'F', 'f');
+wr_snbt(FloatTag) { stream.setf(std::ios::fixed, std::ios::floatfield); stream.setf(std::ios::showpoint); stream << std::setprecision(std::numeric_limits<decltype(value)>::max_digits10) << value << 'f'; }
+
+rd_snbt(DoubleTag) { stream >> value; return !!stream; }
+wr_snbt(DoubleTag) { stream.setf(std::ios::floatfield, std::ios::floatfield); stream.setf(std::ios::showpoint); stream << std::setprecision(std::numeric_limits<decltype(value)>::max_digits10) << value; }
+
+#undef rd_snbt_suffix
+
+#define rd_snbt_typed_array(elementType, typeMarker) \
+  rd_snbt(elementType##ArrayTag) { \
+    skip_whitespaces(stream); \
+    char c; \
+    stream.get(c); \
+    if (c != '[') \
+      return false; \
+    stream.get(c); \
+    if (c != typeMarker) \
+      return false; \
+    stream.get(c); \
+    if (c != ';') \
+      return false; \
+    const TagTypeMask acceptedMask = tagTypeToMask(TagType::elementType); \
+    elementType##Tag numTag; \
+    std::vector<decltype(numTag.value)> elements; \
+    while (stream) { \
+      skip_whitespaces(stream); \
+      if (!stream) \
+        return false; \
+      if (stream.peek() == ']') { \
+        stream.ignore(); \
+        break; \
+      } \
+      if (!(getNextSNBTTagTypes(stream) & acceptedMask)) \
+        return false; \
+      if(!numTag.readSNBT(stream)) \
+        return false; \
+      elements.push_back(numTag.value); \
+      skip_whitespaces(stream); \
+      stream.get(c); \
+      if (c == ']') \
+        break; \
+      if (c != ',') \
+        return false; \
+    } \
+    value.resize(elements.size()); \
+    for (int i = 0; i < elements.size(); ++i) \
+      value.setElement(i, elements[i]); \
+    return true; \
+  }
+
+rd_snbt_typed_array(Byte, 'B')
+wr_snbt(ByteArrayTag) {
+  stream << "[B;";
+  ByteTag numTag;
+  uint32_t count = htonl(value.count);
```

Comment:
Why do you use `htonl` here?

Reply:
> Why do you use `htonl` here?

I just copied this line from here: https://github.com/iRath96/webNBT/blob/1ec291322fad1a751d192dc9094be6b284fcdd28/nbt-utils/nbt_utils.cpp#L211-L214

Reply:
@Vftdan It is used there because NBT requires that integer values are serialized in network (_big endian_) order. That is not necessarily the endianness of the host system (which usually is _little endian_). Hence the conversion from _host to network_ (`hton`).

When you are doing this before your `for`-loop, however, very nasty things happen if the host order and network order mismatch. An array that has only one element (`value.count = 0x00000001`) will cause your local `count` variable to have the value `0x01000000`, causing your loop to run 16777216 times even though there is only one element.